### PR TITLE
linux_testing_bcachefs: 4.16.2018.08.03 -> 4.18.2018.08.31

### LINKED ISSUE #45769: linux_testing_bcachefs broken on commits earlier than 8/21
## Issue description

Older bcachefs kernels were writing invalid dirents which were broken on newer kernels up until recently when the following commit fixed the issue:

https://evilpiepirate.org/git/bcachefs.git/commit/?id=446219cb11af8b6a4c6b837e336bac45f43854c9

### Steps to reproduce

Create a bcachefs filesystem on the older Nixos version from the following link: https://github.com/NixOS/nixpkgs/commit/d043a30d62ad60bc7a1eb3d0f750a5b9c2d98936

Upgrade to more recent one, attempt to mount and use the filesystem. Get tons of errors and broken fsck attempts.

## Technical details

Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the
results.

 - system: `"x86_64-linux"`
 - host os: `Linux 4.15.0, NixOS, 18.03.133174.edd63e05d1e (Impala)`
 - multi-user?: `yes`
 - sandbox: `no`
 - version: `nix-env (Nix) 2.0.4`
 - channels(root): `"nixos-18.03"`
 - channels(mitch): `"nixos"`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos/nixpkgs`

### PR DESCRIPTION
**Pull request by ghost**
###### Motivation for this change
Fixes #45769

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

**Comment by Mic92** [comment]
@GrahamcOfBorg build bcachefs-tools

**Comment by Mic92** [comment]
Let's see if we can make it in time:

```
@GrahamcOfBorg build linux_testing_bcachefs
```

**Comment by Mic92** [comment]
@Chiiruno please disable aarch64 support for this kernel:

```
extraMeta.platforms = [ "x86_64-linux" ];
```

**Comment by Chiiruno** [comment]
@Mic92 Done.

**Comment by vcunat** [comment]
Well, the kernel builds on my `x86_64-linux`, but this is more about using it with the FS :-)

**Comment by Chiiruno** [comment]
Yes, for me too.
Does nix have a way to streamline filesystem tests in a VM?
Is the build bot overloaded right now by any chance?

**Comment by Mic92** [comment]
It is not overloaded probably but there is a one hour limit, this is ok to fail. I can build locally.

**Comment by Mic92** [comment]
We have tests for the installer and zfs in nixpkgs/nixos/tests. Bcachefs can be added as well.

**Comment by vcunat** [comment]
For _this_ merge I'm fine if the FS now works for you &ndash; you did test that, right?  PR with a reasonably written VM test would surely get merged, too.

**Comment by Chiiruno** [comment]
@vcunat I only tested compilation and what was checked off in the OP.
I'm not savvy enough from nix vm to do that, unfortunately.
I have, however, tested bcachefs before, and it's worked, so I would assume this one does too, but (I hope) this probably means very little.

**Comment by Chiiruno** [comment]
If you want a VM test, it'll have to wait until tomorrow, but chances are you could merge this now and it would be fine.

**Comment by Mic92** [comment]
I currently cannot test this either.

**Comment by Mic92** [comment]
The build also works on my machine.

**Comment by vcunat** [comment]
Then I'd wait until someone runs the FS with this, on bare metal or VM.

**Comment by Chiiruno** [comment]
My apologies, I stopped using bcachefs after I lost a lot of data, ironically due to the error this commit supposedly fixes. I'll see what I can do about tests, and at the very least I'll run this in a standard VM now that I have the chance. (Is there a way to easily add an empty disk to a nixos vm, though?)

**Comment by Chiiruno** [comment]
@vcunat First time making a test, that was surprisingly easy.
Seems to work, this can be merged I think.

File: nixos/tests/bcachefs.nix
Comment (Mic92) [inline nixos/tests/bcachefs.nix]: if the test is supposed to be run by hydra, it should be also referenced in `nixpkgs/nixos/release.nix`

**Comment by Chiiruno** [comment]
@Mic92 All good?

**Review by vcunat — APPROVED** [review]

**Pull request merged**
